app/services/rag_wise/discussion_extraction_service.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+class RagWise::DiscussionExtractionService
+  def initialize(course, topic, posts)
+    @course = course
+    @topic = topic
+    @posts = Course::Discussion::Post.where(id: posts.pluck(:id))
+  end
+
+  def call
+    {
+      topic_title: sanitized_topic_title,
+      discussion: formatted_discussion
+    }
+  end
+
+  private
+
+  def sanitized_topic_title
+    sanitize_text(@topic[:title])
+  end
+
+  def formatted_discussion
+    @posts.filter_map do |post|
+      sanitized_text = sanitize_text(post[:text])
+
+      {
+        role: post_creator_role(@course, post),
+        text: sanitized_text
+      }.tap do |hash|
+        hash[:image_captions] = image_captions(post) if post.attachments
+      end
+    end
+  end
+
+  def sanitize_text(text)
+    ActionController::Base.helpers.strip_tags(text)
+  end
+
+  def image_captions(post)
+    llm_service = RagWise::LlmService.new
+
+    post.attachments.map do |attachment|
```

Comment:
Does `@posts` includes the `:attachments` as well?  
Otherwise, there's another N+1 here, as well as L30.

Reply:
used includes(:creator, :attachment_references)